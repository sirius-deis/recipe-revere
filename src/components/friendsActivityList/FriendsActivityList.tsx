import { FC, useEffect, useRef, useState } from "react";
import { useQuery } from "@apollo/client";
import Loader from "../loader/Loader";
import ErrorBox from "../errorBox/ErrorBox";
import FriendsActivity from "../friendsActivity/FriendsActivity";
import useOnScreen from "../../hooks/useOnScreen";
import List from "../list/List";
import { GET_FRIENDS_ACTIVITY } from "../../queries/queries";

type FriendsActivityType = {
  _id: string;
  user: {
    _id: string;
    name: string;
    pictures: string;
  };
  activity: string;
  date: number;
};

const FriendsActivityList: FC = () => {
  const firstFetch = useRef(true);
  const [page, setPage] = useState(0);
  const [friendsActivity, setFriendsActivity] = useState<
    FriendsActivityType[] | []
  >([]);
  const { loading, error } = useQuery(GET_FRIENDS_ACTIVITY, {
    variables: { page },
    onCompleted: (data) => {
      firstFetch.current &&= false;
      setFriendsActivity((currentState) => {
        return [...currentState, ...data];
      });
    },
  });
  const ref = useRef<HTMLDivElement>(null);
  const isIntersecting = useOnScreen(ref);
  useEffect(() => {
    setPage((prevState: number) => {
      return prevState + 1;
    });
  }, [isIntersecting]);
  if (loading && firstFetch.current) {
    return <Loader />;
  }
  if (error) {
    return <ErrorBox message={error.message} />;
  }

  return (
    <section>
      <List borderBottom>
        {friendsActivity.map(
          (
            { _id, user: { _id: userId, name, pictures }, activity, date },
            index
          ) => {
            return (
              <FriendsActivity
                key={_id}
                _id={userId}
                name={name}
                picture={pictures[0]}
                activity={activity}
                time={date}
                ref={friendsActivity.length - 1 <= index ? null : ref}
              />
            );
          }
        )}
      </List>
      {loading && !firstFetch.current && <Loader />}
    </section>
  );
};

export default FriendsActivityList;
