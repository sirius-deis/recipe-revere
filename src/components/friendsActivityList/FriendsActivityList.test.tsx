import { render, screen } from "@testing-library/react";
import FriendsActivityList from "./FriendsActivityList";

const intersectionObserverMock = () => ({
  observe: () => null,
  disconnect: () => null
})
window.IntersectionObserver = jest.fn().mockImplementation(intersectionObserverMock);

jest.mock("@apollo/client", () => ({
  ...jest.requireActual("@apollo/client"),
  useQuery: jest.fn()
}))

import { useQuery } from "@apollo/client";

jest.mock("../loader/Loader", () => () => <div>Loading...</div>)

describe("FriendsActivityList", () => {
  const mockData = [{
    _id: "12",
    user: {
      _id: "34",
      name: "John",
      pictures: "https://example.com/pic.jpg"
    },
    activity: "Created a new post",
    date: 1633072800000
  }, {
    _id: "23",
    user: {
      _id: "45",
      name: "Sam",
      pictures: "https://example.com/pic.jpg"
    },
    activity: "Liked post",
    date: 1633076800000
  }]
  it("should match snapshot", () => {
    (useQuery as jest.Mock).mockReturnValue({
      loading: false,
      error: null,
      data: mockData
    })
    const { container } = render(<FriendsActivityList />)
    expect(container).toMatchSnapshot();
  })
  it("should render loading state", () => {
    (useQuery as jest.Mock).mockReturnValue({
      loading: true
    })
    render(<FriendsActivityList />)
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  })
  it("should render error state", () => {
    const errorMessage = "Error fetching friends activity";
    (useQuery as jest.Mock).mockReturnValue({
      loading: false,
      error: new Error(errorMessage)
    })
    render(<FriendsActivityList />)
    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  })
  it("should render no activity message", () => {
    (useQuery as jest.Mock).mockReturnValue({
      loading: false,
      error: null,
      data: { friendsActivity: [] }
    })
    render(<FriendsActivityList />)
    expect(screen.getByText("No friends activity yet")).toBeInTheDocument();
  })
})