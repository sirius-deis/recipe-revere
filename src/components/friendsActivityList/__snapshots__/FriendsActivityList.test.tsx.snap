// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FriendsActivityList should match snapshot 1`] = `
<div>
  <p>
    No friends activity yet
  </p>
</div>
`;
